### PR TITLE
Update installation documentation for Symfony Flex

### PR DESCRIPTION
**Pull request by kunicmarko20**
I am targeting this branch, because I am updating documentation for 4.x

## Subject

I did a fresh install of SonataUser and docs weren't really ready for use with the new structure of Symfony Flex. Basically, with current documentation, someone that has never used Sonata packages would have a problem getting SonataUser working.

### CODE REVIEW

**Comment by kunicmarko20** [comment]
There are some problems I encountered while installing and I need your suggestion on how to proceed and should this be added to docs:

**1. When installing SonataAdmin there is an error:**

```
The service "sonata.admin.block.admin_list" has a dependency on a non-existent service "templating".  
```

Solution is:

```
framework:
    templating:
        engines: ['twig']
```

As there are PRs on all bundles and Sonata is moving away from templating, I am not sure if this should be added?

**2. When installing SonataUser there is an error:**

`friendsofsymfony/user-bundle v2.0.2 requires symfony/form ^2.7 || ^3.0
`

There still is no release of FOSUserBundle.

Solution:

`composer require friendsofsymfony/user-bundle:2.1.x-dev`

**3. After installing FOSUser like I did in 2. because there is no recipe the bundle is added to `bundles.php` but cache clear after the installation throws an error because there is no configuration for FOSUser:**

```
In ArrayNode.php line 238:
                                                                     
The child node "db_driver" at path "fos_user" must be configured.  
```

Solution:

Add configuration for FOSUser, I mean we know what to do and error is clear, but I am thinking about people that are installing this, this is connected to 2. with recipe in contrib repo this goes away.

**4. The last error I encountered was this issue https://github.com/sonata-project/SonataUserBundle/issues/942#issuecomment-342126397 and that comment is the solution by @jordisala1991** 

Did not get into details so not sure about that one at all I just added that config and it worked.

**_Next things are not issues but questions:_**

**Do we still need the mention all the bundles in `Enable the bundle` part or just sonata-user?**

```
php composer require sonata-project/user-bundle --no-update
php composer require sonata-project/doctrine-orm-admin-bundle  --no-update # optional
php composer require friendsofsymfony/rest-bundle  --no-update # optional when using api
php composer require nelmio/api-doc-bundle  --no-update # optional when using api
php composer require sonata-project/google-authenticator --no-update  # optional
php composer update
```

`sonata-project/doctrine-orm-admin-bundle` this one will probably already be installed as you have to follow sonata admin installation first.

**Should I add `Sonata block configuration` to part `2.3` and move** 
```
sonata_block:
    blocks:
        #...
        sonata.user.block.menu:    # used to display the menu in profile pages
        sonata.user.block.account: # used to display menu option (login option)
        sonata.block.service.text: # used to if you plan to use Sonata user routes
```
from the ACL configuration to that new part?


This is it, for now, will wait for your feedback and hopefully this gets merged soon so people can install it without problems.

**Comment by greg0ire** [comment]
@covex-nn please answer 1.

File: docs/reference/installation.rst [hunk 1]
```diff
@@ -25,42 +26,43 @@ Enable the Bundle
...
+        Sonata\UserBundle\SonataUserBundle::class => ['all' => true],
+    ];
 
 Configuration
 -------------
-When using ACL, the ``UserBundle`` can prevent `normal` user to change settings of `super-admin` users, to enable this add to the configuration:
+
+ACL Configuration
+~~~~~~~~~~~~~~~~~
+When using ACL, the ``UserBundle`` can prevent `normal` user to change
```
Comment (greg0ire) [inline docs/reference/installation.rst]: users

File: docs/reference/installation.rst [hunk 2]
```diff
@@ -25,42 +26,43 @@ Enable the Bundle
...
+settings of `super-admin` users, to enable this add to the configuration:
```
Comment (greg0ire) [inline docs/reference/installation.rst]: "use the following configuration"

File: docs/reference/installation.rst [hunk 3]
```diff
@@ -25,42 +26,43 @@ Enable the Bundle
 
 .. code-block:: bash
 
-    php composer.phar require sonata-project/user-bundle --no-update
-    php composer.phar require sonata-project/doctrine-orm-admin-bundle  --no-update # optional
-    php composer.phar require friendsofsymfony/rest-bundle  --no-update # optional when using api
-    php composer.phar require nelmio/api-doc-bundle  --no-update # optional when using api
-    php composer.phar require sonata-project/google-authenticator --no-update  # optional
-    php composer.phar update
+    php composer require sonata-project/user-bundle --no-update
```
Comment (core23) [inline docs/reference/installation.rst]: You could remove `php` as well

**Comment by covex-nn** [comment]
> The service "sonata.admin.block.admin_list" has a dependency on a non-existent service "templating".  

@kunicmarko20 this error will go away when https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/794 and https://github.com/sonata-project/SonataAdminBundle/pull/4897 will be released (newest BlockBundle does not use `templating` service). Then a recipe for `sonata-project/admin-bundle` must be created and aproved. See https://github.com/symfony/recipes-contrib/pull/264/files (it is almost ready, i believe that `sonata_admin.title` should be removed from there).

With a recipe AdminBundle could be installed and configured by this commands:

```
composer create-project symfony/skeleton .
composer config extra.symfony.allow-contrib true
composer require sonata-project/doctrine-orm-admin-bundle
```

Also, for `sonata-project/doctrine-phpcr-admin-bundle` https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/517 must be released.

**Comment by covex-nn** [comment]
Also with https://github.com/FriendsOfSymfony/FOSUserBundle/pull/2708 and https://github.com/FriendsOfSymfony/FOSUserBundle/pull/2707 it will be possible to create a recipe for FOSUserBundle

**Comment by kunicmarko20** [comment]
Thank you for info @covex-nn. So I won't add 1., 2. and 3. to docs. The only issue that is left is number 4. and my question about `Enable the bundle` part and sonata block config.

**Comment by covex-nn** [comment]
@kunicmarko20 i'm waiting for a PRs reviews in FOSUserBundle already 15 days =) and may be we will wait a month (or even more) for a `2.1.0` release. So, may be a few words about FOSUserBundle configuration would be fine.

Any way here is how FOSUserBundle could be configured for Flex https://github.com/FriendsOfSymfony/FOSUserBundle/issues/2704#issuecomment-356740957. The main part is:

```yaml
fos_user:
    service:
        mailer: fos_user.mailer.twig_swift
```

This parameter will not require `templating` service.

File: docs/reference/installation.rst [hunk 4]
```diff
@@ -7,7 +7,8 @@ Installation
 Prerequisites
 -------------
 
-PHP 7 and Symfony 2.8 or 3 are needed to make this bundle work; there are also some Sonata dependencies that need to be installed and configured beforehand:
+PHP 7 and Symfony 2.8 or 3 are needed to make this bundle work, there are
```
Comment (jordisala1991) [inline docs/reference/installation.rst]: 7.1. Symfony 2.8, >=3.2 or 4 right?

File: docs/reference/installation.rst [hunk 5]
```diff
@@ -25,42 +26,43 @@ Enable the Bundle
...
-    php composer.phar update
+    php composer require sonata-project/user-bundle --no-update
+    php composer require sonata-project/doctrine-orm-admin-bundle  --no-update # optional
+    php composer require friendsofsymfony/rest-bundle  --no-update # optional when using api
+    php composer require nelmio/api-doc-bundle  --no-update # optional when using api
+    php composer require sonata-project/google-authenticator --no-update  # optional
+    php composer update
 
-Next, be sure to enable the bundles in your and ``AppKernel.php`` file:
+Next, be sure to enable the bundles in your ``bundles.php`` file if they
```
Comment (jordisala1991) [inline docs/reference/installation.rst]: shouldn't we explain for sf 3 and sf 4?

on Sf 2/3 it should be AppKernel

File: docs/reference/installation.rst (hunk 4 above)
Comment (kunicmarko20) [reply docs/reference/installation.rst]: 3.2 support is ending next week, should I mark it as 3.3 instead?

File: docs/reference/installation.rst (hunk 5 above)
Comment (kunicmarko20) [reply docs/reference/installation.rst]: As flex is now the default way of creating new projects I though that should only be documented or am I wrong? @greg0ire I remember we talked about this on slack WDYT?

Comment (greg0ire) [reply docs/reference/installation.rst]: We MUST assume everyone is using sf 4 and flex. We CAN add sf 2/3 instructions.

Comment (kunicmarko20) [reply docs/reference/installation.rst]: That means I would need to add a note or something like `but if you are using symfony 3, do it this way`? so SHOULD I?

Comment (greg0ire) [reply docs/reference/installation.rst]: It's really up to you. Personally, I think people that don't know how to register a bundle and also don't start with sf 4 might be quite rare, so it's probably not needed.

Comment (kunicmarko20) [reply docs/reference/installation.rst]: I agree with that because current Symfony doc points to 4.0 and if you are using 3.4 you probably went with flex, if not then you know what you are doing. I do not mind adding docs for 3. If rest of the people think it should be added.

**Comment by kunicmarko20** [comment]
I added docs for non-flex but I do not really like all the notes with `If you are not using Symfony Flex`, what do you think?

File: docs/reference/installation.rst [hunk 6]
```diff
@@ -25,14 +26,35 @@ Enable the Bundle
...
+
+    // config/bundles.php
+
+    return [
+        //...
+        Sonata\AdminBundle\SonataAdminBundle::class => ['all' => true],
+        Sonata\CoreBundle\SonataCoreBundle::class => ['all' => true],
+        Sonata\BlockBundle\SonataBlockBundle::class => ['all' => true],
+        Sonata\EasyExtendsBundle\SonataEasyExtendsBundle::class => ['all' => true],
+        FOS\UserBundle\FOSUserBundl::class => ['all' => true],
```
Comment (kunicmarko20) [inline docs/reference/installation.rst]: `Bundle`

File: docs/reference/installation.rst [hunk 7]
```diff
@@ -223,65 +296,118 @@ The last part is to define 3 new access control rules:
...
 .. code-block:: bash
 
-    php app/console sonata:easy-extends:generate SonataUserBundle -d src
+    bin/console sonata:easy-extends:generate SonataUserBundle --dest=src --namespace_prefix=App
+
+.. note::
+    If you are not using Symfony Flex, use command without ``--namespace_prefix=App``.
 
-If you specify no parameter, the files are generated in ``app/Application/SonataUserBundle`` but you can specify the path with ``--dest=src``
+With provided parameters, the files are generated in ``src/Application/SonataUserBundle``.
```
Comment (kunicmarko20) [inline docs/reference/installation.rst]: This is wrong, it should be `src/Application/Sonata/UserBundle`

File: docs/reference/installation.rst [hunk 8]
```diff
@@ -42,38 +64,42 @@ Next, be sure to enable the bundles in your and ``AppKernel.php`` file:
...
-            new Sonata\AdminBundle\SonataAdminBundle(),
-            new Sonata\CoreBundle\SonataCoreBundle(),
-            new Sonata\BlockBundle\SonataBlockBundle(),
-            new Sonata\EasyExtendsBundle\SonataEasyExtendsBundle(),
-            // ...
-            new FOS\UserBundle\FOSUserBundle(),
-            new Sonata\UserBundle\SonataUserBundle(),
-            // ...
-        );
+         return array(
```
Comment (jordisala1991) [inline docs/reference/installation.rst]: Why did you add an extra space?

Comment (kunicmarko20) [reply docs/reference/installation.rst]: Didn't see it, will be fixed.

Comment (OskarStark) [reply docs/reference/installation.rst]: and you can use short array syntax here, too

File: docs/reference/installation.rst [hunk 9]
```diff
@@ -83,29 +109,33 @@ When using ACL, the ``UserBundle`` can prevent `normal` user to change settings
...
-    fos_user:
-        db_driver:      orm # can be orm or odm
-        firewall_name:  main
-        user_class:     Sonata\UserBundle\Entity\BaseUser
+    sonata_block:
+        blocks:
+            #...
+            sonata.user.block.menu:    # used to display the menu in profile pages
+            sonata.user.block.account: # used to display menu option (login option)
+            sonata.block.service.text: # used to if you plan to use Sonata user routes
```
Comment (OskarStark) [inline docs/reference/installation.rst]: the comment looks strange

**Comment by greg0ire** [comment]
Thanks @kunicmarko20 !
